src/System.Net.Security/src/System/PinnableBufferCache.cs

```diff
@@ -0,0 +1,552 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System
+{
+    internal sealed class PinnableBufferCache
+    {
+        /// <summary>
+        /// Create a new cache for pinned byte[] buffers.
+        /// </summary>
+        /// <param name="cacheName">A name used in diagnostic messages</param>
+        /// <param name="numberOfElements">The size of byte[] buffers in the cache (they are all the same size)</param>
+        public PinnableBufferCache(string cacheName, int numberOfElements) : this(cacheName, () => new byte[numberOfElements]) { }
+
+        /// <summary>
+        /// Get a buffer from the buffer manager.  If no buffers exist, allocate a new one.
+        /// </summary>
+        public byte[] AllocateBuffer() { return (byte[])Allocate(); }
+
+        /// <summary>
+        /// Return a buffer back to the buffer manager.
+        /// </summary>
+        public void FreeBuffer(byte[] buffer) { Free(buffer); }
+
+        /// <summary>
+        /// Create a PinnableBufferCache that works on any object (it is intended for OverlappedData)
+        /// </summary>
+        internal PinnableBufferCache(string cacheName, Func<object> factory)
```

Comment:
Why `Func<object>`? Things will go badly if it doesn't return a `byte[]`, so `Func<byte[]>` seems more apt...